src/niri.rs

```diff
@@ -3728,47 +3736,162 @@ impl Niri {
         }
     }
 
+    pub fn render_for_screencopy_with_damage(
+        &mut self,
+        renderer: &mut GlesRenderer,
+        output: &Output,
+    ) {
+        let _span = tracy_client::span!("Niri::render_for_screencopy_with_damage");
+
+        let mut screencopy_state = mem::take(&mut self.screencopy_state);
+        let elements = OnceCell::new();
+
+        for queue in screencopy_state.queues_mut() {
+            let (damage_tracker, screencopy) = queue.split();
+            if let Some(screencopy) = screencopy {
+                if screencopy.output() == output {
+                    let elements = elements.get_or_init(|| {
+                        self.render(renderer, output, true, RenderTarget::ScreenCapture)
+                    });
+                    // FIXME: skip elements if not including pointers
+                    let render_result = Self::render_for_screencopy_internal(
+                        renderer,
+                        output,
+                        elements,
+                        true,
+                        damage_tracker,
+                        screencopy,
+                    );
+                    match render_result {
+                        Ok(damages) => {
+                            if let Some(damages) = damages {
+                                screencopy.damage(damages);
+                                queue.pop().submit(false);
+                            } else {
+                                trace!("no damage found, waiting till next redraw");
+                            }
+                        }
+                        Err(err) => {
+                            // recreate damage tracker to report full damage next check.
+                            *damage_tracker =
+                                OutputDamageTracker::new((0, 0), 1.0, Transform::Normal);
+                            queue.pop();
+                            warn!("{err}");
+                        }
+                    }
+                };
+            }
+        }
+
+        self.screencopy_state = screencopy_state;
+    }
+
     pub fn render_for_screencopy(
```

Comment:
Since this handles only non-damage screencopies, maybe rename this to something like `render_for_screencopy_no_damage`